java/WordCountOptimized.java

```diff
@@ -217,33 +207,66 @@ public static int tokenizeAndSubmitBlock(TObjectIntCustomHashMap<byte[]> counts,
         }
         return 0;
     }
-    
-    private static void submitWord(TObjectIntCustomHashMap<byte[]> m, byte[] word) {
-        m.adjustOrPutValue(word, 1, 1);        
+
+    private static final int TABLE_SIZE = 1024 * 1024;
+    private static final int TABLE_MASK = TABLE_SIZE - 1;
+
+    static class WordMap {
+        final byte[][] keys = new byte[TABLE_SIZE][];
+        final int[] values = new int[TABLE_SIZE];
+        int size = 0;
     }
-    
+
+    private static int submitWord(byte[][] keys, int[] values, int size,
+                                   byte[] word, int off, int len) {
+        int hashCode = (int) XX_HASH.hashBytes(word, off, len);
+        int index = hashCode & TABLE_MASK;
```

Comment:
The ability to handle arbitrary input datasets should be considered part of the acceptance criteria for a submission, since the input corpus may change.   See also discussion here about overly large 'words' (another issue): https://github.com/juditacs/wordcount/pull/68